src/SelfRoleManager.ts

```diff
@@ -272,6 +281,25 @@ export class SelfRoleManager extends EventEmitter {
         break;
     }
   }
+
+  #listenRoleChanges(rolesToEmojis: RoleToEmojiData[]) {
+    this.client.on(Events.GuildMemberUpdate, async (oldMember, newMember) => {
+      const rolesToRemove = rolesToEmojis.filter(
+        (rte) =>
+          rte.dependencies?.some((dep) => {
+            const role = dep instanceof Role ? dep.id : dep;
+            return oldMember.roles.cache.has(role) && !newMember.roles.cache.has(role);
+          }) ?? false
+      );
+      for (const { role } of rolesToRemove) {
+        const roleToRemove = role instanceof Role ? role.id : role;
+        if (newMember.roles.cache.has(roleToRemove)) {
+          await removeRole(newMember, roleToRemove);
+          this.emit(SelfRoleManagerEvents.roleRemove, roleToRemove, newMember);
```

Comment:
As you can see above in the switch, I actually check whether the member update performed well. In fact, many issues could arise while you try to interact with a guild member (in this context, you try to remove a role from a member).
A check on the returned value of `removeRole` should be done to prevent the broadcast of the `roleRemove` event in case of a false-positive!

Also, you emit such event... But according to the documentation, there are **3 parameters** to it: 
1. the guild member
2. the role
3. **the user action**

At line 298, you ignored the third one. You cannot just ignore a parameter!! This would be considered introducing a bug for people who are listening to the event and acting on it.
You also changed the type of the returned values (the role is supposed to be a `Role` object). This would also be considered a bug...

Reply:
As we don't have an userAction for the automatic role remove. I suggest creating a new event: "automaticRoleRemove". 
```ts
/**
 * Emitted when the user wants a role but does not have the required roles to apply for it.
 * @event SelfRoleManager#automaticRoleRemove
 * @param {Role} role
 * @param {GuildMember} member
 * @example
 * manager.on(SelfRoleManagerEvents.automaticRoleRemove, (role, member) => {});
 */
```

Or, we could make the UserAction parameter optional to detect if there was a userAction to remove a role.

The first suggestion clearly separates both cases, but the second allows using a single event when a role is removed from a user. 
The second option could be considered as a breaking change while the first is only a new event and not break any existing use, but it will not detect automatic removal of roles. Thus, both will introduce changes...
What do you think?

Reply:
After some thought on the matter, I guess it would be best to make `userAction` nullable and introduce a breaking change in the next version. Could you create a new file under the `types`  directory named `UserAction.ts` which would export the following the type please?
```ts
export type UserAction = ButtonInteraction | MessageReaction | PartialMessageReaction | null;
```

You should then be able to edit all the variables "userAction" to use that type. Then, you should emit the event `roleRemove` with the guild member and the role object as well as a `null` value for the third parameter.

Doing this means we change the current API. Could you also go in `package.json` and update the major version please? We will be on 3.0.0 with this.

Reply:
I like this option.
I will do that!